src/jitter_rng.rs

```diff
@@ -0,0 +1,591 @@
+    // added when accessing memory. The memory size should be larger than the L1
+    // caches as outlined in the documentation and the associated testing.
+    //
+    // The L1 cache has a very high bandwidth, albeit its access rate is usually
+    // slower than accessing CPU registers. Therefore, L1 accesses only add minimal
+    // variations as the CPU has hardly to wait. Starting with L2, significant
+    // variations are added because L2 typically does not belong to the CPU any more
+    // and therefore a wider range of CPU wait states is necessary for accesses.
+    // L3 and real memory accesses have even a wider range of wait states. However,
+    // to reliably access either L3 or memory, the ec->mem memory must be quite
+    // large which is usually not desirable.
+    #[inline(never)]
+    fn memaccess(&mut self) {
+        let mut index = self.mem_prev_index;
+        for _ in 0..(self.random_loop_cnt(7) + 128) {
+            // Addition of memblocksize - 1 to index with wrap around logic to
+            // ensure that every memory location is hit evenly.
+            // The modulus also allows the compiler to remove the indexing
+            // bounds check.
+            index = (index + MEMORY_BLOCKSIZE - 1) % MEMORY_SIZE;
+
+            // memory access: just add 1 to one byte
+            // memory access implies read from and write to memory location
+            let tmp = self.mem[index];
+            self.mem[index] = tmp.wrapping_add(1);
+        }
+        self.mem_prev_index = index;
+    }
+
+
+    // Stuck test by checking the:
+    // - 1st derivation of the jitter measurement (time delta)
+    // - 2nd derivation of the jitter measurement (delta of time deltas)
+    // - 3rd derivation of the jitter measurement (delta of delta of time deltas)
+    //
+    // All values must always be non-zero.
+    // This test is a heuristic to see whether the last measurement holds entropy.
+    fn stuck(&mut self, current_delta: i64) -> bool {
+        let delta2 = self.last_delta - current_delta;
+        let delta3 = delta2 - self.last_delta2;
+
+        self.last_delta = current_delta;
+        self.last_delta2 = delta2;
+
+        current_delta == 0 || delta2 == 0 || delta3 == 0
+    }
+
+    // This is the heart of the entropy generation: calculate time deltas and
+    // use the CPU jitter in the time deltas. The jitter is injected into the
+    // entropy pool.
+    //
+    // Ensure that `self.prev_time` is primed before using the output of this
+    // function. This can be done by calling this function and not using its
+    // result.
+    fn measure_jitter(&mut self) -> Option<()> {
+        // Invoke one noise source before time measurement to add variations
+        self.memaccess();
+
+        // Get time stamp and calculate time delta to previous
+        // invocation to measure the timing variations
+        let time = (self.timer)();
+        let current_delta = time.wrapping_sub(self.prev_time) as i64;

```

Comment:
It might be worth adding a note that wrapping arithmetic on unsigned integers followed by cast to i64 generates correct deltas in most cases (including where both times have the highest bit set and where one has wrapped). [The exception is where the actual delta is not representable as an `i64`, but this shouldn't be the case.]